cpp/src/groupby/sort/result_cache.hpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/detail/aggregation.hpp>
+
+#include <unordered_map>
+
+namespace cudf {
+namespace experimental {
+namespace groupby {
+namespace detail {
+
+struct aggregation_equality {
+  struct typed_aggregation_equality
+  {
+    std::shared_ptr<aggregation> const& lhs;
+    std::shared_ptr<aggregation> const& rhs;
+
+    template <aggregation::Kind k>
+    bool operator()() const {
+      using agg_type = experimental::detail::kind_to_type<k>;
+      auto typed_lhs = static_cast<agg_type const*>(lhs.get());
+      auto typed_rhs = static_cast<agg_type const*>(rhs.get());
+      return *typed_lhs == *typed_rhs;
+    }
+  };
+
+  bool operator()(std::shared_ptr<aggregation> const& lhs,
+                  std::shared_ptr<aggregation> const& rhs) const
+  {
+    if (lhs->kind != rhs->kind)
+      return false;
+    
+    return experimental::detail::aggregation_dispatcher(lhs->kind,
+              typed_aggregation_equality{lhs, rhs});
+  }
+};
+
+struct aggregation_hash {
+  size_t operator()(std::shared_ptr<aggregation> const& key) const noexcept {
+    if (key) {
+      return key->kind;
+    } else {
+      return 0;
+    }
+  }
+};
+
```

Comment:
Would like to see a comment/description of the purpose and intended usage for this class.

Reply:
The work from this PR has been merged as part of #3597 . I'll try to remember to make this change in my upcoming hash groupby PR